pkg/build/prometheus/metrics.go

```diff
@@ -0,0 +1,172 @@
+		BuildSubsystem+separator+ActiveBuildCount,
+		"Show current counts of active builds by namespace, name, phase, and with start time noted",
+		[]string{"namespace", "name", "phase", "startime"},
+		nil,
+	)
+	descBuildDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildDuration,
+		"Histogram of durations in seconds of terminated builds, by termination type",
+		[]string{"namespace", "name", "phase"},
+		nil,
+	)
+	descStageDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+StageDuration,
+		"Histogram of durations in seconds of stages that have finished, by stage name",
+		[]string{"namespace", "name", "stage"},
+		nil,
+	)
+	descStepDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+StepDuration,
+		"Histogram of durations in seconds of steps that have finished, by stage name and step name",
+		[]string{"namespace", "name", "stage", "step"},
+		nil,
+	)
+	bc         = buildCollector{}
+	registered = false
+	new        = strings.ToLower(string(buildapi.BuildPhaseNew))
+	pending    = strings.ToLower(string(buildapi.BuildPhasePending))
+	running    = strings.ToLower(string(buildapi.BuildPhaseRunning))
+)
+
+type buildCollector struct {
+	lister internalversion.BuildLister
+}
+
+// kube-state-metrics leverages prometheus.Registerer but does not exist in the version of prometheus
+// vendored into origin
+
+func IntializeMetricsCollector(buildLister internalversion.BuildLister) {
+	bc.lister = buildLister
+	// unit tests unearthed duplicate registration with prometheus via multiple calls to new build controller
+	if !registered {
+		prometheus.MustRegister(&bc)
+		registered = true
+	}
+	glog.V(4).Info("build metrics registered with prometheus")
+}
+
+// Describe implements the prometheus.Collector interface.
+func (bc *buildCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- descTerminalBuildCount
+	ch <- descActiveBuildCount
+	ch <- descBuildDuration
+	ch <- descStageDuration
+	ch <- descStepDuration
+}
+
+// Collect implements the prometheus.Collector interface.
+func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
+	result, err := bc.lister.List(kselector.Everything())
+
+	if err != nil {
+		glog.V(4).Infof("Collect err %v", err)
+		return
+	}
+
+	for _, b := range result {
+		bc.collectBuild(ch, b)
+	}
+
+	glog.V(4).Infof("reconcileActiveBuildMetrics looked at %d builds", len(result))
+
+}
+
+func (bc *buildCollector) collectBuild(ch chan<- prometheus.Metric, b *buildapi.Build) {
+	addGauge := func(desc *prometheus.Desc, v float64, lv ...string) {
+		lv = append([]string{b.ObjectMeta.Namespace, b.ObjectMeta.Name}, lv...)
+		ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, v, lv...)
+	}
+	addHistogram := func(desc *prometheus.Desc, v float64, m map[float64]uint64, lv ...string) {
+		lv = append([]string{b.ObjectMeta.Namespace, b.ObjectMeta.Name}, lv...)
+		ch <- prometheus.MustNewConstHistogram(desc, 1, v, m, lv...)
+	}
+
+	creation := b.ObjectMeta.CreationTimestamp.String()
```

Comment:
It's unlikely we need to know this to the precision of the exact value, and if so, it should be sampled as a start time, and we are already calculating within the approximate range of the sample.

The "correct" way to track starts is to do a metric whose value is the number of seconds/milliseconds since the epoch.  And then you have to label it to be unique, which creates a high cardinality outcome anyway.  We'll catch most of this data at lower resolution from the other scraping, so it's just not that important.